DbgPack/pack2.py

```diff
@@ -1,133 +1,187 @@
+        """
+
+        :param assets: List of assets to export
+        :param name: name of file to export to
+        :param outdir: path to save file
+        """
+
+        makedirs(outdir, exist_ok=True)
+
+        with BinaryStructWriter(outdir / name) as writer:
+            data_size = sum([x.size for x in assets])
+
+            writer.write(_MAGIC)
+            writer.uint32LE(len(assets))
+            writer.uint64LE(0)  # Overwrite this later
+            writer.uint64LE(data_size + 0x200)
+            writer.uint64LE(256)
+
+            # Padding
+            while writer.tell() < data_size + 0x200:
+                writer.write(b'\x00')
+
+            data_offset = 0x200
+            for a in sorted(assets, key=lambda x: x.name_hash if isinstance(x, Asset2) else crc64(x.name)):
+                if isinstance(a, (Asset2, LooseAsset)):
+                    writer.uint64LE(a.name_hash)
+                else:
+                    writer.uint64LE(crc64(a.name))
+
+                writer.uint64LE(data_offset)
+                writer.uint64LE(a.size)
+                writer.uint32LE(0x10)  # Compression flag
+                writer.uint32LE(a.crc32)  # PTS doesn't care if the checksums don't match
 
-    def __init__(self, path: str, namelist: List[str] = None):
+                # Write data
+                writer.write_to(a.data, data_offset)
+                data_offset += a.size
+
+            pack_size = writer.tell()
+            writer.seek(0x8, 0)
+            writer.uint64LE(pack_size)
+
+    def __init__(self, path: Path, namelist: List[str] = None):
         super().__init__(path)
-        self.path = path
         self._namelist = namelist
-        with BinaryStructReader(path) as reader:
-            magic = reader.read(4)
-            assert magic == b'PAK\x01'
+
+        with BinaryStructReader(self.path) as reader:
+            assert reader.read(len(_MAGIC)) == _MAGIC, 'invalid pack2 magic'
             self.asset_count = reader.uint32LE()
-            self.file_size = reader.uint64LE()
+            self.size = reader.uint64LE()
             self.map_offset = reader.uint64LE()
 
             reader.seek(self.map_offset)
             self.raw_assets = {}
             for i in range(self.asset_count):
                 name_hash = reader.uint64LE()
+
                 offset = reader.uint64LE()
-                size = reader.uint64LE()
-                zipped_flag = reader.uint32LE()
+                zipped_size = reader.uint64LE()
+                zip_flag = reader.uint32LE()
                 crc32 = reader.uint32LE()
-                asset = Asset(name_hash=name_hash, crc32=crc32, offset=offset, length=size, path=self.path)
-                self.raw_assets.update({int(asset.name_hash): asset})
 
-        self.assets = {}
-        self._update_asset_names(namelist)
+                # HACK: This could probably be contained in a function for use in Asset2.data()
```

Comment:
Go ahead and move the logic for determining the zipped and unzipped size into `Asset2`